src/map.c

```diff
@@ -664,6 +664,49 @@ int wally_map_keypath_add(struct wally_map *map_in,
     return ret;
 }
 
+int wally_map_tap_keypath_add(struct wally_map *hashes_map_in,
+                          struct wally_map *paths_map_in,
+                          const unsigned char *pub_key, size_t pub_key_len,
+                          const unsigned char *tapleaf_hashes, size_t tapleaf_hashes_len,
+                          const unsigned char *fingerprint, size_t fingerprint_len,
+                          const uint32_t *path, size_t path_len)
+{
+    unsigned char *value;
+    size_t value_len, i;
+    int ret;
+
+    if (!hashes_map_in || !paths_map_in || !fingerprint || fingerprint_len != BIP32_KEY_FINGERPRINT_LEN ||
+        BYTES_INVALID(path, path_len) || pub_key_len != EC_XONLY_PUBLIC_KEY_LEN)
+        return WALLY_EINVAL;
+
+	/* TODO add hashes_map_in hash verification */
+    if (paths_map_in->verify_fn != wally_keypath_xonly_public_key_verify)
+        return WALLY_EINVAL; /* Not a taproot keypath map */
+
+	/* TODO take the tapleaf hashes value instead of copy */
```

Comment:
You can remove this, wally always needs internal ownership of held data and so will always take a copy.

Reply:
ah ok, `wally_map_keypath_add` seems to clear out the caller's field, I'll remove

Reply:
done

Reply:
It makes a copy which it swaps in, so the action is atomic.